seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala

```diff
@@ -0,0 +1,110 @@
+
+import java.util.concurrent.TimeoutException
+
+import scala.annotation.tailrec
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
+import scala.util.Try
+
+class SeleniumComJSRunner(browserProvider: SeleniumBrowser,
+    libs: Seq[ResolvedJSDependency], code: VirtualJSFile, keepAlive: Boolean)
+    extends SeleniumAsyncJSRunner(browserProvider, libs, code, keepAlive)
+    with ComJSRunner {
+
+  protected def envName: String =
+    "SeleniumComJSRunner on " + browserProvider.name
+
+  def send(msg: String): Unit = {
+    awaitForBrowser()
+    browser.getWebDriver.executeScript(
+        "this.scalajsSeleniumComJSRunnerChannel.recvMessage(arguments[0])", msg)
+    browser.processConsoleLogs(console)
+  }
+
+  def receive(timeout: Duration): String = {
+    awaitForBrowser(timeout)
+    @tailrec def loop(): String = {
+      val script = "return this.scalajsSeleniumComJSRunnerChannel.popOutMsg();"
+      browser.getWebDriver.executeScript(script) match {
+        case null =>
+          loop()
+
+        case msg: String =>
+          browser.processConsoleLogs(console)
+          msg
+
+        case obj =>
+          // Here we only try to get the console because it uses the same
+          // communication channel that is potentially corrupted.
+          Try(browser.processConsoleLogs(console))
+          BrowserDriver.illFormattedScriptResult(obj)
+      }
+    }
+    loop()
+  }
+
+  def close(): Unit = {
+    browser.processConsoleLogs(console)
+    if (!keepAlive)
+      browser.close()
+  }
+
+  override protected def initFiles(): Seq[VirtualJSFile] =
+    super.initFiles() :+ comSetupFile()
+
+  protected def comSetupFile(): VirtualJSFile = {
+    val code = {
+      s"""
+         |(function() {
+         |  // Buffers for outgoing messages
+         |  var sendMsgBufIn = []; // push to this one
+         |  var sendMsgBufOut = []; // pop from this one
+         |
+         |  var onReceive = null;
+         |
+         |  function closeCom() {
+         |    if (sendMsgBufIn.length + sendMsgBufOut.length != 0) {
+         |      setTimeout(closeCom, 10);
+         |    } else {
+         |      seleniumAsyncJSRunnerClose = true;
+         |    }
+         |  };
+         |
+         |  this.scalajsCom = {
+         |    init: function(recvCB) {
+         |      onReceive = recvCB;
+         |    },
+         |    send: function(msg) {
+         |      sendMsgBufIn.push(msg);
+         |    },
+         |    close: closeCom
+         |  }
+         |
+         |  this.scalajsSeleniumComJSRunnerChannel = {
+         |    popOutMsg: function() {
+         |      if (sendMsgBufOut.length == 0) {
+         |         sendMsgBufOut = sendMsgBufIn.reverse();
+         |         sendMsgBufIn = [];
+         |      }
+         |      return sendMsgBufOut.pop();
+         |    },
+         |    recvMessage: function(msg) {
+         |      onReceive(msg);
+         |    }
+         |  }
+         |})();
```

Comment:
`}).call(this);`
Otherwise `this` is `undefined` inside the lambda in strict mode.